{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio import SwissProt\n",
    "from urllib.request import urlopen\n",
    "from Bio.Align.Applications import ClustalwCommandline, MuscleCommandline\n",
    "from Bio import AlignIO\n",
    "from Bio import Phylo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('AYTRYVDNNFCGPDGYPLECIKDLLARAGKASCTLSEQLDFIDTKRGVYCCREH...KGG'), id='YP_009725298.1', name='YP_009725298.1', description='YP_009725298.1 nsp2 [Severe acute respiratory syndrome coronavirus 2]', dbxrefs=[])"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nsp2 = SeqIO.read(open(\"C:/Users/Zé Freitas/Desktop/Mestrado/Labs_Bioinf/Trabalho prático/scripts/Labs_Bioinf/NSP2/nsp2.fasta\"),\"fasta\")\n",
    "nsp2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('AYTRYVDNNFCGPDGYPLECIKDLLARAGKASCTLSEQLDFIDTKRGVYCCREH...KGG'), id='sp|P0DTC1|181-818', name='sp|P0DTC1|181-818', description='sp|P0DTC1|181-818', dbxrefs=[])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Abrir ficheiro SwissProt localmente FASTA\n",
    "ficheiro = SeqIO.read(open(\"C:/Users/Zé Freitas/Desktop/Mestrado/Labs_Bioinf/Trabalho prático/scripts/Labs_Bioinf/NSP2/M20210105A94466D2655679D1FD8953E075198DA82BA1B88.fasta\"),\"fasta\")\n",
    "ficheiro "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RecName: Full=Replicase polyprotein 1a; Short=pp1a; AltName: Full=ORF1a polyprotein; Contains: RecName: Full=Host translation inhibitor nsp1; AltName: Full=Leader protein; AltName: Full=Non-structural protein 1; Short=nsp1; Contains: RecName: Full=Non-structural protein 2; Short=nsp2; AltName: Full=p65 homolog; Contains: RecName: Full=Non-structural protein 3; Short=nsp3; EC=3.4.19.12 {ECO:0000269|PubMed:32726803}; EC=3.4.22.-; AltName: Full=PL2-PRO; AltName: Full=Papain-like protease {ECO:0000303|PubMed:32726803}; AltName: Full=Papain-like proteinase; Short=PL-PRO; Contains: RecName: Full=Non-structural protein 4; Short=nsp4; Contains: RecName: Full=3C-like proteinase; Short=3CL-PRO; Short=3CLp; EC=3.4.22.69; AltName: Full=Main protease; Short=Mpro {ECO:0000303|PubMed:32272481}; AltName: Full=Non-structural protein 5; Short=nsp5; AltName: Full=SARS coronavirus main proteinase; Contains: RecName: Full=Non-structural protein 6; Short=nsp6; Contains: RecName: Full=Non-structural protein 7; Short=nsp7; Contains: RecName: Full=Non-structural protein 8; Short=nsp8; Contains: RecName: Full=Non-structural protein 9; Short=nsp9; Contains: RecName: Full=Non-structural protein 10; Short=nsp10; AltName: Full=Growth factor-like peptide; Short=GFL; Contains: RecName: Full=Non-structural protein 11; Short=nsp11;\n",
      "[<Bio.SwissProt.Reference object at 0x000001DEC05A3040>, <Bio.SwissProt.Reference object at 0x000001DEC05A7340>, <Bio.SwissProt.Reference object at 0x000001DEC05A7580>, <Bio.SwissProt.Reference object at 0x000001DEC05A7310>, <Bio.SwissProt.Reference object at 0x000001DEC059D490>, <Bio.SwissProt.Reference object at 0x000001DEC1CB13D0>, <Bio.SwissProt.Reference object at 0x000001DEC1CB1310>, <Bio.SwissProt.Reference object at 0x000001DEC1CB1340>, <Bio.SwissProt.Reference object at 0x000001DEC1CB1370>, <Bio.SwissProt.Reference object at 0x000001DEC1CB13A0>, <Bio.SwissProt.Reference object at 0x000001DEC1CB1400>, <Bio.SwissProt.Reference object at 0x000001DEC1CB1430>, <Bio.SwissProt.Reference object at 0x000001DEC1CB1460>]\n",
      "authors: Wu F., Zhao S., Yu B., Chen Y.-M., Wang W., Song Z.-G., Hu Y., Tao Z.-W., Tian J.-H., Pei Y.-Y., Yuan M.-L., Zhang Y.-L., Dai F.-H., Liu Y., Wang Q.-M., Zheng J.-J., Xu L., Holmes E.C., Zhang Y.-Z.\n",
      "title: A new coronavirus associated with human respiratory disease in China.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Liu X., Zhang B., Jin Z., Yang H., Rao Z.\n",
      "title: The crystal structure of 2019-nCoV main protease in complex with an inhibitor N3.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Zhang L., Lin D., Sun X., Curth U., Drosten C., Sauerhering L., Becker S., Rox K., Hilgenfeld R.\n",
      "title: Crystal structure of SARS-CoV-2 main protease provides a basis for design of improved alpha-ketoamide inhibitors.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Jin Z., Du X., Xu Y., Deng Y., Liu M., Zhao Y., Zhang B., Li X., Zhang L., Peng C., Duan Y., Yu J., Wang L., Yang K., Liu F., Jiang R., Yang X., You T., Liu X., Yang X., Bai F., Liu H., Liu X., Guddat L.W., Xu W., Xiao G., Qin C., Shi Z., Jiang H., Rao Z., Yang H.\n",
      "title: Structure of Mpro from COVID-19 virus and discovery of its inhibitors.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Gao Y., Yan L., Huang Y., Liu F., Zhao Y., Cao L., Wang T., Sun Q., Ming Z., Zhang L., Ge J., Zheng L., Zhang Y., Wang H., Zhu Y., Zhu C., Hu T., Hua T., Zhang B., Yang X., Li J., Yang H., Liu Z., Xu W., Guddat L.W., Wang Q., Lou Z., Rao Z.\n",
      "title: Structure of the RNA-dependent RNA polymerase from COVID-19 virus.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Yin W., Mao C., Luan X., Shen D.D., Shen Q., Su H., Wang X., Zhou F., Zhao W., Gao M., Chang S., Xie Y.C., Tian G., Jiang H.W., Tao S.C., Shen J., Jiang Y., Jiang H., Xu Y., Zhang S., Zhang Y., Xu H.E.\n",
      "title: Structural basis for inhibition of the RNA-dependent RNA polymerase from SARS-CoV-2 by remdesivir.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Dai W., Zhang B., Su H., Li J., Zhao Y., Xie X., Jin Z., Liu F., Li C., Li Y., Bai F., Wang H., Cheng X., Cen X., Hu S., Yang X., Wang J., Liu X., Xiao G., Jiang H., Rao Z., Zhang L.K., Xu Y., Yang H., Liu H.\n",
      "title: Structure-based design of antiviral drug candidates targeting the SARS-CoV-2 main protease.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Hillen H.S., Kokic G., Farnung L., Dienemann C., Tegunov D., Cramer P.\n",
      "title: Structure of replicating SARS-CoV-2 polymerase.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Frick D.N., Virdi R.S., Vuksanovic N., Dahal N., Silvaggi N.R.\n",
      "title: Molecular Basis for ADP-ribose Binding to the Mac1 Domain of SARS-CoV-2 Nsp3.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Wang Q., Wu J., Wang H., Gao Y., Liu Q., Mu A., Ji W., Yan L., Zhu Y., Zhu C., Fang X., Yang X., Huang Y., Gao H., Liu F., Ge J., Sun Q., Yang X., Xu W., Liu Z., Yang H., Lou Z., Jiang B., Guddat L.W., Gong P., Rao Z.\n",
      "title: Structural Basis for RNA Replication by the SARS-CoV-2 Polymerase.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Shin D., Mukherjee R., Grewe D., Bojkova D., Baek K., Bhattacharya A., Schulz L., Widera M., Mehdipour A.R., Tascher G., Geurink P.P., Wilhelm A., van der Heden van Noort G.J., Ovaa H., Mueller S., Knobeloch K.P., Rajalingam K., Schulman B.A., Cinatl J., Hummer G., Ciesek S., Dikic I.\n",
      "title: Papain-like protease regulates SARS-CoV-2 viral spread and innate immunity.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Thoms M., Buschauer R., Ameismeier M., Koepke L., Denk T., Hirschenberger M., Kratzat H., Hayn M., Mackens-Kiani T., Cheng J., Straub J.H., Stuerzel C.M., Froehlich T., Berninghausen O., Becker T., Kirchhoff F., Sparrer K.M.J., Beckmann R.\n",
      "title: Structural basis for translational shutdown and immune evasion by the Nsp1 protein of SARS-CoV-2.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "authors: Lei X., Dong X., Ma R., Wang W., Xiao X., Tian Z., Wang C., Wang Y., Li L., Ren L., Guo F., Zhao Z., Zhou Z., Xiang Z., Wang J.\n",
      "title: Activation and evasion of type I interferon responses by SARS-CoV-2.\n",
      "['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n"
     ]
    }
   ],
   "source": [
    "#Abrir ficheiro SwissProt localmente full infor\n",
    "ficheiro = open(\"C:/Users/Zé Freitas/Desktop/Mestrado/Labs_Bioinf/Trabalho prático/scripts/Labs_Bioinf/NSP2/uniprot-yourlist_M20210102E5A08BB0B2D1C45B0C7BC3B55FD265560A91A6U.txt\")\n",
    "ficheiro\n",
    "\n",
    "record = SwissProt.read(ficheiro)\n",
    "print(record.description)\n",
    "print(record.references)\n",
    "\n",
    "for ref in record.references:\n",
    "    print(\"authors:\", ref.authors)\n",
    "    print(\"title:\", ref.title)\n",
    "    print(record.organism_classification)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RecName: Full=Replicase polyprotein 1a; Short=pp1a; AltName: Full=ORF1a polyprotein; Contains: RecName: Full=Host translation inhibitor nsp1; AltName: Full=Leader protein; AltName: Full=Non-structural protein 1; Short=nsp1; Contains: RecName: Full=Non-structural protein 2; Short=nsp2; AltName: Full=p65 homolog; Contains: RecName: Full=Non-structural protein 3; Short=nsp3; EC=3.4.19.12 {ECO:0000269|PubMed:32726803}; EC=3.4.22.-; AltName: Full=PL2-PRO; AltName: Full=Papain-like protease {ECO:0000303|PubMed:32726803}; AltName: Full=Papain-like proteinase; Short=PL-PRO; Contains: RecName: Full=Non-structural protein 4; Short=nsp4; Contains: RecName: Full=3C-like proteinase; Short=3CL-PRO; Short=3CLp; EC=3.4.22.69; AltName: Full=Main protease; Short=Mpro {ECO:0000303|PubMed:32272481}; AltName: Full=Non-structural protein 5; Short=nsp5; AltName: Full=SARS coronavirus main proteinase; Contains: RecName: Full=Non-structural protein 6; Short=nsp6; Contains: RecName: Full=Non-structural protein 7; Short=nsp7; Contains: RecName: Full=Non-structural protein 8; Short=nsp8; Contains: RecName: Full=Non-structural protein 9; Short=nsp9; Contains: RecName: Full=Non-structural protein 10; Short=nsp10; AltName: Full=Growth factor-like peptide; Short=GFL; Contains: RecName: Full=Non-structural protein 11; Short=nsp11;\n",
      "[<Bio.SwissProt.Reference object at 0x000001DEC059DA60>, <Bio.SwissProt.Reference object at 0x000001DEC1CB1100>, <Bio.SwissProt.Reference object at 0x000001DEC1CB1280>, <Bio.SwissProt.Reference object at 0x000001DEC1CB10A0>, <Bio.SwissProt.Reference object at 0x000001DEC1CE50D0>, <Bio.SwissProt.Reference object at 0x000001DEC1CE5040>, <Bio.SwissProt.Reference object at 0x000001DEC1CE5760>, <Bio.SwissProt.Reference object at 0x000001DEC1CE5820>, <Bio.SwissProt.Reference object at 0x000001DEC1CE57F0>, <Bio.SwissProt.Reference object at 0x000001DEC1CE5850>, <Bio.SwissProt.Reference object at 0x000001DEC1CE5910>, <Bio.SwissProt.Reference object at 0x000001DEC1CE5940>, <Bio.SwissProt.Reference object at 0x000001DEC1CE5970>]\n"
     ]
    }
   ],
   "source": [
    "#Abrir ficheiro SwissProt pela internet\n",
    "url = \"https://www.uniprot.org/uniprot/P0DTC1.txt\"\n",
    "handle = urlopen(url)\n",
    "handle \n",
    "\n",
    "recorde = SwissProt.read(handle)\n",
    "recorde\n",
    "print(recorde.description)\n",
    "print(recorde.references)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#BLAST de proteinas \n",
    "blast = NCBIWWW.qblast(\"blastp\", \"nr\", nsp2.format(\"fasta\"))\n",
    "\n",
    "save_file = open(\"my_blast_nsp2.xml\",\"w\")\n",
    "save_file.write(blast.read())\n",
    "save_file.close()\n",
    "blast.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "****Alignment****\n",
      "sequence: gb|QOS45331.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QQG74758.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPI19453.1| ORF1a polyprotein, partial [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4378\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPZ45818.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QQE72107.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPN54778.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPF48271.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF48295.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF48403.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF48439.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF48583.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOU89401.1| ORF1a polyprotein, partial [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4366\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QNS28380.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOF18748.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOU95653.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4402\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOL72912.1| ORF1a polyprotein, partial [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4386\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPJ71341.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPN00870.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QNN96407.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOW16764.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOW16776.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOW78874.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOX06568.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOX06580.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPP19251.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOL76308.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOL21562.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPF56221.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QQE69361.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOZ00431.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOH26417.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOH26429.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOQ71956.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOW78778.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOW78790.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPN00690.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPF48619.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF48631.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF48643.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF48655.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF48679.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPK51128.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPJ70837.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QNN26455.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPN54586.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPF56233.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOD07140.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QNT10059.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QNT10107.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOL21622.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOL21658.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF15979.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPF54649.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF54661.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF54673.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF54685.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QPF54709.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOI09374.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOL71640.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOL73896.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPJ71017.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOU88297.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPF50065.1| ORF1a polyprotein, partial [Severe acute respiratory syndrome coronavirus 2] >gb|QPF50137.1| ORF1a polyprotein, partial [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4381\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPF50077.1| ORF1a polyprotein, partial [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4380\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPN04177.1| ORF1a polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 4405\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOE78176.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 7096\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOZ00406.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 7096\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOY99289.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 7096\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOU82346.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 7096\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOW17091.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 7096\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPD98106.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 7096\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPF50724.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 7096\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QPF49050.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 7096\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QNN95315.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 7096\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOL76103.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOL76211.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOL76427.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOL76739.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2] >gb|QOL76823.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 7096\n",
      "e value: 0.0\n",
      "****Alignment****\n",
      "sequence: gb|QOU88908.1| ORF1ab polyprotein [Severe acute respiratory syndrome coronavirus 2]\n",
      "length: 7096\n",
      "e value: 0.0\n"
     ]
    }
   ],
   "source": [
    "#Leitura dos resultados do BLASTp\n",
    "result = open(\"my_blast_nsp2.xml\")\n",
    "blast_records = NCBIXML.parse(result)\n",
    "blast_records\n",
    "\n",
    "E_VALUE_THRESH = 0.05\n",
    "for blast_record in blast_records:\n",
    "    for alignment in blast_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "                if hsp.expect < E_VALUE_THRESH:\n",
    "                    print ('****Alignment****')\n",
    "                    print ('sequence:', alignment.title)\n",
    "                    print ('length:', alignment.length)\n",
    "                    print ('e value:', hsp.expect)\n",
    "result.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class ClustalwCommandline in module Bio.Align.Applications._Clustalw:\n",
      "\n",
      "class ClustalwCommandline(Bio.Application.AbstractCommandline)\n",
      " |  ClustalwCommandline(cmd='clustalw', **kwargs)\n",
      " |  \n",
      " |  Command line wrapper for clustalw (version one or two).\n",
      " |  \n",
      " |  http://www.clustal.org/\n",
      " |  \n",
      " |  Notes\n",
      " |  -----\n",
      " |  Last checked against versions: 1.83 and 2.1\n",
      " |  \n",
      " |  References\n",
      " |  ----------\n",
      " |  Larkin MA, Blackshields G, Brown NP, Chenna R, McGettigan PA,\n",
      " |  McWilliam H, Valentin F, Wallace IM, Wilm A, Lopez R, Thompson JD,\n",
      " |  Gibson TJ, Higgins DG. (2007). Clustal W and Clustal X version 2.0.\n",
      " |  Bioinformatics, 23, 2947-2948.\n",
      " |  \n",
      " |  Examples\n",
      " |  --------\n",
      " |  >>> from Bio.Align.Applications import ClustalwCommandline\n",
      " |  >>> in_file = \"unaligned.fasta\"\n",
      " |  >>> clustalw_cline = ClustalwCommandline(\"clustalw2\", infile=in_file)\n",
      " |  >>> print(clustalw_cline)\n",
      " |  clustalw2 -infile=unaligned.fasta\n",
      " |  \n",
      " |  You would typically run the command line with clustalw_cline() or via\n",
      " |  the Python subprocess module, as described in the Biopython tutorial.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      ClustalwCommandline\n",
      " |      Bio.Application.AbstractCommandline\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, cmd='clustalw', **kwargs)\n",
      " |      Initialize the class.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __call__(self, stdin=None, stdout=True, stderr=True, cwd=None, env=None)\n",
      " |      Execute command, wait for it to finish, return (stdout, stderr).\n",
      " |      \n",
      " |      Runs the command line tool and waits for it to finish. If it returns\n",
      " |      a non-zero error level, an exception is raised. Otherwise two strings\n",
      " |      are returned containing stdout and stderr.\n",
      " |      \n",
      " |      The optional stdin argument should be a string of data which will be\n",
      " |      passed to the tool as standard input.\n",
      " |      \n",
      " |      The optional stdout and stderr argument may be filenames (string),\n",
      " |      but otherwise are treated as a booleans, and control if the output\n",
      " |      should be captured as strings (True, default), or ignored by sending\n",
      " |      it to /dev/null to avoid wasting memory (False). If sent to a file\n",
      " |      or ignored, then empty string(s) are returned.\n",
      " |      \n",
      " |      The optional cwd argument is a string giving the working directory\n",
      " |      to run the command from. See Python's subprocess module documentation\n",
      " |      for more details.\n",
      " |      \n",
      " |      The optional env argument is a dictionary setting the environment\n",
      " |      variables to be used in the new process. By default the current\n",
      " |      process' environment variables are used. See Python's subprocess\n",
      " |      module documentation for more details.\n",
      " |      \n",
      " |      Default example usage::\n",
      " |      \n",
      " |          from Bio.Emboss.Applications import WaterCommandline\n",
      " |          water_cmd = WaterCommandline(gapopen=10, gapextend=0.5,\n",
      " |                                       stdout=True, auto=True,\n",
      " |                                       asequence=\"a.fasta\", bsequence=\"b.fasta\")\n",
      " |          print(\"About to run: %s\" % water_cmd)\n",
      " |          std_output, err_output = water_cmd()\n",
      " |      \n",
      " |      This functionality is similar to subprocess.check_output(). In general\n",
      " |      if you require more control over running the command, use subprocess\n",
      " |      directly.\n",
      " |      \n",
      " |      When the program called returns a non-zero error level, a custom\n",
      " |      ApplicationError exception is raised. This includes any stdout and\n",
      " |      stderr strings captured as attributes of the exception object, since\n",
      " |      they may be useful for diagnosing what went wrong.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a representation of the command line object for debugging.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> cline\n",
      " |      WaterCommandline(cmd='water', outfile='temp_water.txt', asequence='asis:ACCCGGGCGCGGT', bsequence='asis:ACCCGAGCGCGGT', gapopen=10, gapextend=0.5)\n",
      " |  \n",
      " |  __setattr__(self, name, value)\n",
      " |      Set attribute name to value (PRIVATE).\n",
      " |      \n",
      " |      This code implements a workaround for a user interface issue.\n",
      " |      Without this __setattr__ attribute-based assignment of parameters\n",
      " |      will silently accept invalid parameters, leading to known instances\n",
      " |      of the user assuming that parameters for the application are set,\n",
      " |      when they are not.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5, stdout=True)\n",
      " |      >>> cline.asequence = \"a.fasta\"\n",
      " |      >>> cline.bsequence = \"b.fasta\"\n",
      " |      >>> cline.csequence = \"c.fasta\"\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      ValueError: Option name csequence was not found.\n",
      " |      >>> print(cline)\n",
      " |      water -stdout -asequence=a.fasta -bsequence=b.fasta -gapopen=10 -gapextend=0.5\n",
      " |      \n",
      " |      This workaround uses a whitelist of object attributes, and sets the\n",
      " |      object attribute list as normal, for these.  Other attributes are\n",
      " |      assumed to be parameters, and passed to the self.set_parameter method\n",
      " |      for validation and assignment.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Make the commandline string with the currently set options.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> str(cline)\n",
      " |      'water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5'\n",
      " |  \n",
      " |  set_parameter(self, name, value=None)\n",
      " |      Set a commandline option for a program (OBSOLETE).\n",
      " |      \n",
      " |      Every parameter is available via a property and as a named\n",
      " |      keyword when creating the instance. Using either of these is\n",
      " |      preferred to this legacy set_parameter method which is now\n",
      " |      OBSOLETE, and likely to be DEPRECATED and later REMOVED in\n",
      " |      future releases.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  parameters = None\n",
      "\n",
      "clustalw2 -infile=\"C:/Users/Zé Freitas/Desktop/Mestrado/Labs_Bioinf/Trabalho prático/scripts/Labs_Bioinf/NSP2/NSP2_MA.fasta\"\n",
      "muscle -in \"C:/Users/Zé Freitas/Desktop/Mestrado/Labs_Bioinf/Trabalho prático/scripts/Labs_Bioinf/NSP2/NSP2_MA.fasta\" -out NSP2align.aln -clw\n"
     ]
    }
   ],
   "source": [
    "#Métodos de alinhamento multiplo e criação da árvore filogenética\n",
    "help(ClustalwCommandline)\n",
    "cline = ClustalwCommandline(\"clustalw2\", infile=\"C:/Users/Zé Freitas/Desktop/Mestrado/Labs_Bioinf/Trabalho prático/scripts/Labs_Bioinf/NSP2/NSP2_MA.fasta\")\n",
    "print(cline)\n",
    "\n",
    "cline = MuscleCommandline(input=\"C:/Users/Zé Freitas/Desktop/Mestrado/Labs_Bioinf/Trabalho prático/scripts/Labs_Bioinf/NSP2/NSP2_MA.fasta\", out=\"NSP2align.aln\", clw=True)\n",
    "print(cline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment with 8 rows and 7250 columns\n",
      "--------------------------------------------...--- YP_009725298.1\n",
      "--MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEALTEA...VNN QQM18896.1\n",
      "--MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEAISEA...VNN QIG55944.1\n",
      "--MESLVLGVNEKTHVQLSLPVLQVRDVLVRGFGDSVEEALSEA...--- ALK02469.1\n",
      "--MESLVLGVNEKTHVQLSLPVLQVRDVLVRGFGDSVEEALSEA...VNN ARI44798.1\n",
      "--MESLVLGVNEKTHVQLSLPVLQVRDVLVRGFGDSVEEALAEA...VNN QDF43814.1\n",
      "MSSATGEGSQGARATYRAALNNEKRSDHVALTVPCCGTGAKVTA...--- QCC20712.1\n",
      "MSSVAGVVTQGARNMYRAALNNEKRQDHVSLTTPLCGAGDLALR...--- AWH65953.1\n"
     ]
    }
   ],
   "source": [
    "#Leitura de ficheiro do alinhamento multiplo\n",
    "alignment = AlignIO.read(\"C:/Users/Zé Freitas/Desktop/Mestrado/Labs_Bioinf/Trabalho prático/scripts/Labs_Bioinf/NSP2/NSP2alinhados.fasta\", \"fasta\")\n",
    "print(alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tree(rooted=False, weight=1.0)\n",
      "    Clade()\n",
      "        Clade(branch_length=0.06018)\n",
      "            Clade(branch_length=0.08603)\n",
      "                Clade(branch_length=0.04171)\n",
      "                    Clade(branch_length=0.07664, name='YP_009725298.1')\n",
      "                    Clade(branch_length=-0.03902, name='QQM18896.1')\n",
      "                Clade(branch_length=-0.01167, name='QIG55944.1')\n",
      "            Clade(branch_length=0.46862)\n",
      "                Clade(branch_length=0.22197, name='QCC20712.1')\n",
      "                Clade(branch_length=0.22444, name='AWH65953.1')\n",
      "        Clade(branch_length=0.02511)\n",
      "            Clade(branch_length=0.01496, name='ALK02469.1')\n",
      "            Clade(branch_length=-0.00401, name='ARI44798.1')\n",
      "        Clade(branch_length=0.01768, name='QDF43814.1')\n"
     ]
    }
   ],
   "source": [
    "#Leitura do ficheiro da árvore filogenética\n",
    "arvore = Phylo.read(\"C:/Users/Zé Freitas/Desktop/Mestrado/Labs_Bioinf/Trabalho prático/scripts/Labs_Bioinf/NSP2/NSP2arvore.dnd\", \"newick\")\n",
    "print(arvore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 _____ YP_009725298.1\n",
      "              __|\n",
      "      _______|  | QQM18896.1\n",
      "     |       |\n",
      "  ___|       | QIG55944.1\n",
      " |   |\n",
      " |   |                                       _________________ QCC20712.1\n",
      " |   |______________________________________|\n",
      " |                                          |__________________ AWH65953.1\n",
      "_|\n",
      " | , ALK02469.1\n",
      " |_|\n",
      " | | ARI44798.1\n",
      " |\n",
      " | QDF43814.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Print da árvore filogenética\n",
    "Phylo.draw_ascii(arvore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook NSP2.ipynb to html\n",
      "[NbConvertApp] Writing 308392 bytes to NSP2.html\n"
     ]
    }
   ],
   "source": [
    "!jupyter nbconvert --to html NSP2.ipynb"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
